tools/asdf/src/proto.rs

```diff
@@ -0,0 +1,272 @@
+fn get_versions(proto_temp_dir: VirtualPath) -> FnResult<Vec<String>> {
+    let script_path = clone_repo(proto_temp_dir)?;
+    let script_path = script_path.join("bin").join("list-all").real_path().unwrap().into_os_string().into_string().unwrap();
+
+    let versions = exec_command!("bash", [script_path]).stdout;
+    let versions: Vec<String> = versions.split_whitespace().map(str::to_owned).collect();
+    Ok(versions)
+}
+
+#[plugin_fn]
+pub fn register_tool(Json(input): Json<ToolMetadataInput>) -> FnResult<Json<ToolMetadataOutput>> {    
+    Ok(Json(ToolMetadataOutput {
+        name: input.id,
+        type_of: PluginType::Language,
+        minimum_proto_version: Some(Version::new(0, 45, 2)),
+        default_install_strategy: InstallStrategy::BuildFromSource,
+        plugin_version: Version::parse(env!("CARGO_PKG_VERSION")).ok(),
+		config_schema: Some(schematic::SchemaBuilder::generate::<AsdfPluginConfig>()),
+        ..ToolMetadataOutput::default()
+    }))
+}
+
+#[plugin_fn]
+pub fn build_instructions(
+    Json(input): Json<BuildInstructionsInput>,
+) -> FnResult<Json<BuildInstructionsOutput>> {
+    let env = get_host_environment()?;    
+
+    if env.os.is_windows() {
+        return Err(PluginError::UnsupportedWindowsBuild.into());
+    }
+    
+    let repo = get_repo()?;
+
+    let git = GitSource {
+        url: repo.url.clone(),
+        // Use default branch
+        reference: None,
+        ..Default::default()
+    };
+
+    let mut instructions = Vec::new();
+
+    let version = input.context.version;
+
+    // Set asdf environment variables
+    let install_download_path = &input.context.tool_dir.real_path().unwrap().into_os_string().into_string().unwrap();
+    let cores = if env.os.is_mac() {
+        exec_command!("sysctl -n hw.physicalcpu").stdout
+    } else {
+        exec_command!("nproc").stdout
+    };
+    instructions.append(&mut vec![
+        BuildInstruction::SetEnvVar("ASDF_INSTALL_TYPE".into(), "version".into()),
+        BuildInstruction::SetEnvVar("ASDF_INSTALL_VERSION".into(), version.clone().into()),
+        BuildInstruction::SetEnvVar("ASDF_INSTALL_PATH".into(), install_download_path.clone()),
+        BuildInstruction::SetEnvVar("ASDF_DOWNLOAD_PATH".into(), install_download_path.clone()),
+        BuildInstruction::SetEnvVar("ASDF_CONCURRENCY".into(), cores),
+    ]);
+
+    let id = get_id(None)?;
+
+    // In older versions of asdf there may not be a 'download' script,
+    // instead both download and install were done in the 'install' script.
+    // However, in newer versions, there's two separate 'download' and 'install' scripts.
+    if send_request!(format!("{}/blob/{}/bin/download", repo.url, repo.default_branch)).status == 200 {
+            let download_id = String::from(format!("{id}-{version}-download"));
+            instructions.push(BuildInstruction::InstallBuilder(Box::new(BuilderInstruction {
+                id: download_id.clone(),
+                exe: "bin/download".into(),
```

Comment:
Instead of using 2 builders, you can use 1 builder and use the new `exes` field instead.